# Machine Learning - Scale

Scale Features

When your data has different values, and even different measurement units, it can be difficult to compare them. What is kilograms compared to meters? Or altitude compared to time?

The answer to this problem is scaling. We can scale data into new values that are easier to compare.

Take a look at the table below, it is the same data set that we used in the multiple regression chapter, but this time the volume column contains values in liters instead of cm3 (1.0 instead of 1000).

In [25]:
import pandas as pd
import numpy as np

In [14]:
df = pd.read_csv("cardmodified.csv")

In [18]:
df.head()

Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo     1.0     790   99
1  Mitsubishi  Space Star     1.2    1160   95
2       Skoda      Citigo     1.0     929   95
3        Fiat         500     0.9     865   90
4        Mini      Cooper     1.5    1140  105

In [21]:
df.describe()

Volume       Weight         CO2
count  36.000000    36.000000   36.000000
mean    1.611111  1292.277778  102.027778
std     0.388975   242.123889    7.454571
min     0.900000   790.000000   90.000000
25%     1.475000  1117.250000   97.750000
50%     1.600000  1329.000000   99.000000
75%     2.000000  1418.250000  105.000000
max     2.500000  1746.000000  120.000000

Here the std value of Weight is not correct. We can check it manually by a simple term

In [29]:
weight = df["Weight"]

std_weight = np.std(weight)

print("Standard Deviation :", std_weight)

Standard Deviation : 238.73737443185826


It can be difficult to compare the volume 1.0 with the weight 790, but if we scale them both into comparable values, we can easily see how much one value is compared to the other.

There are different methods for scaling data, in this tutorial we will use a method called standardization.

The standardization method uses this formula:

z = (x - u) / s

Where z is the new value, x is the original value, u is the mean and s is the standard deviation.

If you take the weight column from the data set above, the first value is 790, and the scaled value will be:
(790 - 1292.23) / 238.74 = -2.1

<b>mean of weight is 1292.23</b>

<b>std of weight is 238.74</b>

If you take the volume column from the data set above, the first value is 1.0, and the scaled value will be:

(1.0 - 1.61) / 0.38 = -1.59

Now you can compare -2.07 with -1.59 instead of comparing 790 with 1.0.

We do not have to do this manually, the Python sklearn module has a method called <b>StandardScaler()</b> which returns a Scaler object with methods for transforming data sets.

In [30]:
import pandas as pd
from sklearn import linear_model
from sklearn.preprocessing import StandardScaler

In [31]:
scaler = StandardScaler()

In [33]:
df = pd.read_csv("cardmodified.csv")

In [38]:
X = df[["Volume", "Weight"]]
y = df["CO2"]

In [36]:
scaleX = scaler.fit_transform(X)

In [37]:
scaleX

array([[-1.59336644, -2.10389253],
       [-1.07190106, -0.55407235],
       [-1.59336644, -1.52166278],
       [-1.85409913, -1.78973979],
       [-0.28970299, -0.63784641],
       [-1.59336644, -1.52166278],
       [-0.55043568, -0.76769621],
       [-0.28970299,  0.3046118 ],
       [-0.28970299, -0.7551301 ],
       [-0.0289703 , -0.59595938],
       [-1.33263375, -1.30803892],
       [-0.81116837, -1.26615189],
       [-1.59336644, -0.7551301 ],
       [-0.0289703 , -0.16871166],
       [-0.0289703 ,  0.14125238],
       [-0.0289703 ,  0.15800719],
       [-0.0289703 ,  0.3046118 ],
       [ 1.53542584, -0.05142797],
       [-0.0289703 , -0.72580918],
       [ 1.01396046,  0.14962979],
       [-0.0289703 ,  1.2219378 ],
       [ 1.01396046,  0.5685001 ],
       [ 1.27469315,  0.3046118 ],
       [-0.0289703 ,  0.51404696],
       [ 1.01396046,  0.51404696],
       [-0.28970299,  0.72348212],
       [ 1.01396046,  0.8281997 ],
       [ 1.01396046,  1.81254495],
       [-0.0289703 ,

# Predict CO2 Values

The task in the Multiple Regression chapter was to predict the CO2 emission from a car when you only knew its weight and volume.

When the data set is scaled, you will have to use the scale when you predict values:

In [41]:
regr = linear_model.LinearRegression()
regr.fit(scaleX, y)

LinearRegression()

In [42]:
scaled_ = scaler.transform([[0.2, 1121]])

D:\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [45]:
predictco2 = regr.predict([scaled_[0]])

In [46]:
predictco2

array([89.72038269])

In [48]:
df

Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo     1.0     790   99
1   Mitsubishi  Space Star     1.2    1160   95
2        Skoda      Citigo     1.0     929   95
3         Fiat         500     0.9     865   90
4         Mini      Cooper     1.5    1140  105
5           VW         Up!     1.0     929  105
6        Skoda       Fabia     1.4    1109   90
7     Mercedes     A-Class     1.5    1365   92
8         Ford      Fiesta     1.5    1112   98
9         Audi          A1     1.6    1150   99
10     Hyundai         I20     1.1     980   99
11      Suzuki       Swift     1.3     990  101
12        Ford      Fiesta     1.0    1112   99
13       Honda       Civic     1.6    1252   94
14      Hundai         I30     1.6    1326   97
15        Opel       Astra     1.6    1330   97
16         BMW           1     1.6    1365   99
17       Mazda           3     2.2    1280  104
18       Skoda       Rapid     1.6    1119  104
19        Ford       Focus     2.0    1328  105
20        Ford      Mondeo     1.6    1584   94
21        Opel    Insignia     2.0    1428   99
22    Mercedes     C-Class     2.1    1365   99
23       Skoda     Octavia     1.6    1415   99
24       Volvo         S60     2.0    1415   99
25    Mercedes         CLA     1.5    1465  102
26        Audi          A4     2.0    1490  104
27        Audi          A6     2.0    1725  114
28       Volvo         V70     1.6    1523  109
29         BMW           5     2.0    1705  114
30    Mercedes     E-Class     2.1    1605  115
31       Volvo        XC70     2.0    1746  117
32        Ford       B-Max     1.6    1235  104
33         BMW         216     1.6    1390  108
34        Opel      Zafira     1.6    1405  109
35    Mercedes         SLK     2.5    1395  120